### PR TITLE
Beginning and end of assignment made. Data used in a medium post

### PR DESCRIPTION
**Pull request by JefNtungila**

### CODE REVIEW

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# making_forecasts - "Jephte Ntungila"

# Part 1 - Loading Data
***
**What went well:**

- Good job with importing the provided tools and implementing a good (maybe complex) way of taking in the metrics data from Wikipedia.:

```python

project = 'en.wikipedia'
access = 'all-access'
agent = 'user'
article = 'Veganism'
granularity = 'daily'
start = '20150701'
end = '20190527'

endpoint = f'/metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}'
url = 'https://wikimedia.org/api/rest_v1' + endpoint
response = requests.get(url)
assert response.status_code == 200
df = pd.DataFrame(response.json()['items'])
df = df.rename(columns={'timestamp':'ds', 'views':'y'})
df['ds'] = pd.to_datetime(df['ds'], format='%Y%m%d%H')6570


```

- Good job reading in the weather and setting a variable to parse the dates:
```python

url = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Regression-1/master/module2-evaluating-forecasts/weather-normal-il.csv'
weather = (pd.read_csv(url, parse_dates=['DATE'])
             .rename(columns={'DATE':'ds', 'TMAX':'y'}))

```

**Where you could of improved on:**

-Finish "Do train/test split" section and the rest of the assignment :


# What to work on
- [x] Finish Assignment
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# evaluating_forecasts - "Jephte Ntungila"

# Part 1 - Loading Data/Analysis 
***
**What went well:**

- Good job with importing your pandas data frame from the csv file: 

```python

f = pd.read_csv('https://raw.githubusercontent.com/JefNtungila/Data-story-telling-using-public-data-from-CERN-Run2010B-public-Mu-dataset-/master/Muon_wiki_pageviews-20180501-20190430.csv')
df = df.rename(columns={'Date':'ds', 'Muon':'y'})
```

- Nice use of datetime:

```python


df['ds'] = pd.to_datetime(df['ds'])


```

**Where you could of improved on:**

- Great job splicing: 
```python 

train = df['2015-07-01':'2018-12-31']
test = df['2019-01-01':'2019-05-10']


```

# Part 2 - Graphing 
***
**What went well:**

- Nice use of splicing:

```python

train = uber['2015-07-01':'2019-05-23']


test = uber['2018-05-24': '2019-05-24']



```

- Good showing of through the years .:

```python
ax = train.plot(x='ds', y='y', color='blue', label='Train', figsize=(16,4))
test.plot(x='ds', y='y', color='red', label='Test', ax=ax)
train.plot(x='ds', y='mean_baseline', color='black', 
           label='Mean Baseline, Train', ax=ax)
test.plot(x='ds', y='mean_baseline', color='black', 
          label='Mean Baseline, Test', ax=ax)
plt.title('Average Muon wikipedia page viewers:');


```
- Great use of Prophet library and using the metrics .:

```python
from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(df_cv, metric='mae')

```

# What to work on
- [x] Comment more in code 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# doing_linear_regression- "Jephte Ntungila"

# Part 1 - Loading Data/Analysis findings
***
**What went well:**

- Great job at taking pandas and setting a variable. For future projects naming the main dataset to something like "df" will make it eaiser for yourself and your teammates.: 

```python

unemployment_rate = pd.read_csv('https://raw.githubusercontent.com/JefNtungila/DS-Unit-2-Regression-1/master/module3-doing-linear-regression/unemploymentrate.csv')


unemployment_rate['DATE'] = pd.to_datetime(unemployment_rate['DATE'])


unemployment_rate['DATE'] = pd.DatetimeIndex(unemployment_rate['DATE']).year




```

- Simplistic way of finding target mean. Great job!: 

```python


unemployment_rate = unemployment_rate.groupby(['Year']).mean()


```


- Impressive work on finding Absolute Error and mean from sklearn:

```python

df['Absolute Error'] = df['Error'].abs()
df['Absolute Error'].sum()


df['Absolute Error'].mean()

from sklearn.metrics import mean_absolute_error
mean_absolute_error(y_true = df[target], y_pred = df['Mean Baseline'])


```

**Where you could of improved on:**

- I'd try to make more comments: 


```python 


```

# Part 2 - Graphing 
***
**What went well:**

- Clean and elegant: code. I'm impressed :) :

```python



%matplotlib inline
import seaborn as sns
sns.distplot(df['Incumbent Party Vote Share']);



```

# What to work on
- [x] Comment more in code 
- [x] Less lines and more organized notebooks 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Jephte Ntungila- "Regression 1 Sprint Challenge"

# Part 1. Predict avocado sales
***
**What went well:**

-Good job imporing data and dataset:

```python

import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)



```

## 1.2. Plot the forecast

- Impressive use of forcast and knowledge of Prophet!:
```python


fig1 = m.plot(forecast)

```




```

##1.5. Plot the cross-validation Mean Absolute Percentage Error: metric='mape' 
***
**What went well:**

- Clean and elegant: code. I'm impressed :) :

```python



from fbprophet.plot import plot_cross_validation_metric
fig3 = plot_cross_validation_metric(df_cv, metric='mape')


```

#Stretch Goal (PARTIAL)

# Part 2 

##2.1. Begin with baselines for regression. Calculate the mean absolute error and 𝑅2 score for a mean baseline.

- Went went the base .head: 


``` python


rent.head()





```

**What could be better

-use this syntax for better insight:
```python

rent.describe()

```


## 2.2. Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
-Good job choosing features and getting to it (I added another way as well)):
```python

#Your Code
X = rent[['bedrooms','bathrooms']]
y = rent['price']


features = ['bedrooms','bathrooms']

target = 'price'

X = rent[features]
y = rent[target]

model.fit(X, y)

y_pred = model.predict(X)



train = rent.query('month <= 5')
test  = rent.query('month == 6')
X_train = train[features] 
y_train = train[target]
X_test  = test[features]
y_test  = test[target]


model.fit(X_train, y_train)


y_pred = model.predict(X_test)



#My Syntax

features = ['bedrooms']
target = 'price'
x = rent[features]
y = rent[target]


## Other option

train = rent.query('month < 6')
test = rent.query('month == 6')
train.shape, test.shape

features = ['bedrooms', 'bathrooms']
target = 'price'

```

##2.3. Choose any two features.

-use this syntax for better insight:
```python

```

2.4. Use scikit-learn to fit a Linear Regression model on the train data.

-Not bad for just starting out to use to scikit. Syntax is not bad, but features could be better ( see below ):
```python

from sklearn.metrics import mean_absolute_error, r2_score
import matplotlib.pyplot as plt

def plot_preds(x, y, y_pred):
    plt.scatter(x, y, label='y_true')
    plt.plot(x, y_pred, label='y_pred')
    plt.legend()
    
    
    mae = mean_absolute_error(y, y_pred)
    r2 = r2_score(y, y_pred)
    print('Mean Absolute Error:', mae)
    print('R^2:', r2)
    
    
    
plot_preds(x, y, y_pred)
plt.title('Mean Baseline');

```

**What could be better

-better syntax:


```python

# Choose a class of model by importing the appropriate estimator class from Scikit-Learn.
from sklearn.linear_model import LinearRegression

# Choose model hyperparameters by instantiating this class with desired values.
model = LinearRegression()

# Arrange data into a features matrix and target vector.
X_train = train[features]
y_train = train[target]
X_test = test[features]
y_test = test[target]

X_train.shape, y_train.shape, X_test.shape, y_test.shape

# Fit the model to your data by calling the fit() method of the model instance.
model.fit(X_train, y_train)

# Apply the Model to new data: For supervised learning, often we predict labels for unknown data using the predict() method.
y_pred = model.predict(X_test)

```

2.6. Get regression metrics MAE and 𝑅2 for the test data.

-Great job:
```python


mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print('Test Mean Absolute Error:', mae)
print('Test R^2 Score:', r2)

```

2.7. Get the model's coefficients and intercept.

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

-use this syntax for better insight:
```python
model.intercept_, model.coef_
```

#Stretch Goals_2

```python

import seaborn as sns

target = 'price'
features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
for feature in features:
    sns.lmplot(x=feature, y=target, data=train, scatter_kws={'alpha':0.01})
    
    ```
    
    
    
    ```python
    
    
    def fit(train, test, features):
    X_train = train[features]
    y_train = train[target]
    X_test = test[features]
    y_test = test[target]

    model = LinearRegression()
    model.fit(X_train, y_train)

    print(f'Linear Regression with {len(features)} features: {features}')
    print('Train Mean Absolute Error:', 
          mean_absolute_error(y_train, model.predict(X_train)))
    print('Train R^2 Score:', 
          r2_score(y_train, model.predict(X_train)))
    print('Test Mean Absolute Error:', 
          mean_absolute_error(y_test, model.predict(X_test)))
    print('Test R^2 Score:', 
          r2_score(y_test, model.predict(X_test)))

features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
fit(train, test, features)
```

-with 3 features:

```python
features = ['bedrooms', 'bathrooms', 'longitude']
fit(train, test, features)
```

# What to work on
- [x] Comment more in code 
- [x] Less lines and more organized notebooks 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
[link text](https://)# Jephte Ntungila- Assignment_Random_Forest

# Part 1. Cleaning data

***
**What went well:**

-Nice function to wrangle the data into the test and the X and copy data:

```python
def wrangle(X):
    """Wrangles train, validate, and test sets in the same way"""
    X = X.copy()
    
    # About 3% of the time, latitude has small values near zero,
    # outside Tanzania, so we'll treat these values like zero.
    X['latitude'] = X['latitude'].replace(-2e-08, 0)
    
    # When columns have zeros and shouldn't, they are like null values.
    # So we will replace them with the column mean.
    cols_with_zeros = ['construction_year', 'longitude', 'latitude']
    for col in cols_with_zeros:
        X[col] = X[col].replace(0, np.nan)
        X[col] = X[col].fillna(X[col].mean())
        
    # Convert date_recorded to datetime
    X['date_recorded'] = pd.to_datetime(X['date_recorded'], infer_datetime_format=True)
    
    # Extract year from date_recorded
    X['year_recorded'] = X['date_recorded'].dt.year
    
    # quantity & quantity_group are duplicates, so drop one
    X = X.drop(columns='quantity_group')
    
    # for categoricals with missing values, fill with the category 'MISSING'
    categoricals = X.select_dtypes(exclude='number').columns
    for col in categoricals:
        X[col] = X[col].fillna('MISSING')
    
    return X


train = wrangle(train)
val = wrangle(val)
test = wrangle(test)

```


-Great use of the categorical features analaysis and using the resourses you have at hand:

```python


coefficients = pd.Series(model.coef_[0], X_train_encoded.columns)
plt.figure(figsize=(10,30))
coefficients.sort_values().plot.barh(color='grey');


```


## Logistic regression

- I think that you're use logistic regression is normal and well done. Shorter and more ideal ways could be done. Look at documentation or look at other libraribaes or API's that support logistic logression:
```python

lr = LogisticRegression(solver='lbfgs', multi_class='auto', n_jobs=-1)
lr.fit(train_location, y_train)
print('Logistic Regression')
print('Train Accuracy', lr.score(train_location, y_train))
print('Validation Accuracy', lr.score(val_location, y_val))





```

## Decision Tree 


- Good job at using sklearn and DecisionTreeClassifier. This is commonly used in the data science industry and you should remmber this:

```python
# Documentation 
# https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state=20)
dt.fit(train_location, y_train)
print('Decision Tree')
print('Train Accuracy', dt.score(train_location, y_train))
print('Validation Accuracy', dt.score(val_location, y_val))
```



## Random Forest



**What went well:**

- Clean and elegant code. his is a shorthand for the Pipeline constructor; it does not require, and does not permit, naming the estimators. Instead, their names will be set to the lowercase of their types automatically. :

```python
%%time
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer

pipeline = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True), 
    SimpleImputer(strategy='median'), 
    RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
)

# Fit on train, score on val
pipeline.fit(X_train, y_train)
print('Validation Accuracy', pipeline.score(X_val, y_val))
```


** What could be improved: 

- Could see higher accuracy in the end, but starting art this is is impressieve code:





# What to work on
- [x] Comment more in code 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Jephte Ntungila- Regression 1 Sprint Challenge

# Part 1. Predict avocado sales 

***
**What went well:**

-Good collective use of imports and reading in data of the :

```python
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)

```
#Part 2. Predict NYC apartment rent 

## Prophet Plotting 

- Good job here :

```python

from fbprophet.diagnostics import cross_validation
df_cv = cross_validation(m, initial='730 days', 
                         horizon='90 days', period='45 days') 
df_cv.head()

```

## Calculating MAPE(Mean Absolute Percentage Error)

- Good job at implementing this. while a small line of code, it's very powerful: 

```python
from fbprophet.plot import plot_cross_validation_metric
fig3 = plot_cross_validation_metric(df_cv, metric='mape')
```

# Part 2. Predict NYC apartment rent

##imports

- Nice implementing the dataset into colab: 

```python
import pandas as pd
rent_url = 'https://drive.google.com/uc?export=download&id=1lVVPmh-WYGb_0Gs_lOv22i02-150qDdZ'
rent = pd.read_csv(rent_url)
assert rent.shape == (48295, 7)

```

## MAE

- Excellent work. I mean this because this is important when looking at features:

```python
mean_absolute_error(y_test, y_pred)

```

## R^2 Score

```python
r2_score(y_test, y_pred)
```
# What to work on
- [x] Comment more in code
- [x] Come to PM for questions or help
